Add S3 retry timeout and max attempts config options

elemental-lf opened the pull request:
This is an improved version of #102 and adds new options to change the S3 request timeouts and the maximum number of retries. It will behave differently than before because we're also switching from using the legacy retry mode to standard mode as recommended by the documentation.

wech71 committed:
Feature Issue #100: https://github.com/elemental-lf/benji/issues/100 Storage Module s3: Please add option to increate boto3 retries

elemental-lf committed:
Revamped S3 retry and timeout options

This is based on the initial work of @wech71

elemental-lf commented:
@wech71 could you please review and if possible test these changes?

wech71 reviewed (approved):
Looks good to me :-)

wech71 commented:
I just tried it on a non-critical machine and got

botocore.exceptions.InvalidRetryConfigurationError: Cannot provide retry configuration for "mode". Valid retry configuration options are: 'max_attempts'

It seems this option requires a newer boto3 library

boto3                    1.11.4
botocore                 1.14.4

I think this was added in commit 8d0429c826f1bddef77cfb229640ec642cab5654 on Feb 13. 2020, so the minimum version required for botocore which supports this would be 1.15

Just tested it with a 
   pip install boto3==1.15
now a benji scrub works fine.

=> The setup.py should be updated to reflect this

elemental-lf committed:
Adjust boto3 version requirement

We require a newer version of boto3 now to get the support for configuring the
retry logic mode.

elemental-lf committed:
Update minio test setup to a more recent release of minio

elemental-lf commented:
Thank you for testing this! I've updated the version requirement and will now proceed to merge to master.
